server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -239,23 +239,48 @@ public void onResponse(Releasable connectionReference) {
                     // which point the NodeConnectionsService will have taken ownership of it.
                     registerConnection(destination, connectionReference);
 
-                    pendingJoinInfo.message = PENDING_JOIN_WAITING_RESPONSE;
-                    transportService.sendRequest(
-                        destination,
-                        JOIN_ACTION_NAME,
-                        joinRequest,
-                        TransportRequestOptions.of(null, TransportRequestOptions.Type.PING),
-                        new TransportResponseHandler.Empty() {
+                    // It's possible that our cluster applier is still applying an earlier cluster state (maybe stuck waiting on IO), in
+                    // which case the master will accept our join and add us to the cluster but we won't be able to apply the joining state
+                    // fast enough and will be kicked out of the cluster for lagging, which can happen repeatedly and be a little
+                    // disruptive. To avoid this we send the join from the applier thread which ensures that it's not busy doing something
+                    // else.
+                    pendingJoinInfo.message = PENDING_JOIN_WAITING_APPLIER;
+                    clusterApplier.onNewClusterState(
+                        "joining " + destination.descriptionWithoutAttributes(),
+                        () -> null,
+                        new ActionListener<>() {
                             @Override
-                            public void handleResponse(TransportResponse.Empty response) {
-                                pendingJoinInfo.message = PENDING_JOIN_WAITING_STATE; // only logged if state delayed
-                                pendingOutgoingJoins.remove(dedupKey);
-                                logger.debug("successfully joined {} with {}", destination, joinRequest);
-                                lastFailedJoinAttempt.set(null);
+                            public void onResponse(Void unused) {
+                                pendingJoinInfo.message = PENDING_JOIN_WAITING_RESPONSE;
```

Comment:
Can we assert that we are on the right thread (protecting against spawning the listener call in the future):

```
        assert Thread.currentThread().getName().contains('[' + ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME + ']')
            || Thread.currentThread().getName().startsWith("TEST-") : Thread.currentThread().getName();
```

Reply:
Sure, see 3da642a54699e1934c925e70ef4d45a52c6ace9a. It wouldn't matter much if we forked the listener as long as we did _something_ on the applier thread first - we fork off to a transport thread pretty much straight away anyway. But it's handy documentation at least.